src/fsharp/FSharp.Core/seq.fs

```diff
@@ -1577,57 +1577,56 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Sum")>]
         let inline sum (source: seq< (^a) >) : ^a = 
-            use e = source.GetEnumerator() 
             let mutable acc = LanguagePrimitives.GenericZero< (^a) >
-            while e.MoveNext() do
-                acc <- Checked.(+) acc e.Current
+            for item in source do
+                acc <- Checked.(+) acc item
```

Comment:
> The value type enumerator can't be utilized when the input type is only IEnumerable (a reference type) - it isn't even visible to the compiler. The enumerator will be boxed to IEnumerator But take a look at the IL to check.

C# compiler works like this according to Eric Lippert:

> the first thing we do is check to see if IEnumerable has a method called GetEnumerator. If it does, then we check to see whether the type it returns has method MoveNext and property current. If it does, then the foreach loop is generated entirely using direct calls to those methods and properties. Only if "the pattern" cannot be matched do we fall back to looking for the interfaces.

Even if we get boxed IEnumerable, we could get its original type via `object.GetType()` and check is that type has the GetEnumerator method + MoveNext + Current. As far as I understand, it happens at compile time and not at runtime.